File: Crypto_Exchange/Orderbook/python/6_orderbook_bingx.py
#!/usr/bin/env python3
"""BingX Order Book Feed for CLOB Aggregator"""

import gzip
import json
import time
from datetime import datetime, timezone, timedelta
from threading import Thread
from typing import List, Tuple
from websocket import create_connection


class BingxOrderBook:
    """BingX WebSocket order book feed"""
    
    def __init__(self, aggregator, symbol="BTC-USDT", depth=20):
        self.aggregator = aggregator
        self.symbol = symbol
        self.depth = depth
        self.running = False
        self.url = "wss://open-api-ws.bingx.com/market"
    
    def start(self):
        """Start WebSocket feed in background thread"""
        self.running = True
        Thread(target=self._run, daemon=True, name="bingx").start()
    
    def stop(self):
        """Stop WebSocket feed"""
        self.running = False
    
    def _run(self):
        """Main WebSocket loop with auto-reconnect"""
        while self.running:
            try:
                ws = create_connection(self.url)
                subscribe_msg = {
                    "id": "e745cd6d-d0f6-4a70-8d5a-043e4c741b40",
                    "reqType": "sub",
                    "dataType": f"{self.symbol}@depth20"
                }
                ws.send(json.dumps(subscribe_msg))
                print(f"✓ BingX connected")
                
                while self.running:
                    msg = ws.recv()
                    
                    try:
                        if isinstance(msg, bytes):
                            msg = gzip.decompress(msg).decode('utf-8')
                        data = json.loads(msg)
                    except (gzip.BadGzipFile, UnicodeDecodeError, json.JSONDecodeError):
                        continue
                    
                    if "ping" in data:
                        ws.send(json.dumps({"pong": data["ping"]}))
                        continue
                    
                    if "data" not in data:
                        continue
                    
                    book_data = data["data"]
                    if not isinstance(book_data, dict) or "bids" not in book_data or "asks" not in book_data:
                        continue
                    
                    bids = sorted(((float(p), float(v)) for p, v in book_data["bids"]), 
                                  key=lambda x: x[0], reverse=True)
                    asks = sorted(((float(p), float(v)) for p, v in book_data["asks"]), 
                                  key=lambda x: x[0])
                    
                    # Sanitize crossed levels
                    if bids and asks:
                        bids = [(p, v) for p, v in bids if p < asks[0][0]]
                        asks = [(p, v) for p, v in asks if p > (bids[0][0] if bids else 0)]
                    
                    if bids and asks:
                        self.aggregator.update_exchange('bingx', bids, asks)
                    
            except Exception as e:
                if self.running:
                    print(f"BingX error: {e}, reconnecting...")
                    time.sleep(2)


if __name__ == "__main__":
    # Standalone test mode
    from itertools import accumulate
    
    class MockAggregator:
        def update_exchange(self, name, bids, asks):
            self.bids, self.asks = bids, asks
    
    agg = MockAggregator()
    feed = BingxOrderBook(agg)
    feed.start()
    
    DEPTH = 16
    
    try:
        while True:
            time.sleep(0.5)
            if hasattr(agg, 'bids') and agg.bids and hasattr(agg, 'asks') and agg.asks:
                now_utc = datetime.now(timezone.utc)
                now_gmt8 = now_utc.astimezone(timezone(timedelta(hours=8)))
                
                bids = agg.bids[:DEPTH] + [(0.0, 0.0)] * (DEPTH - len(agg.bids[:DEPTH]))
                asks = agg.asks[:DEPTH] + [(0.0, 0.0)] * (DEPTH - len(agg.asks[:DEPTH]))
                
                spread = asks[0][0] - bids[0][0]
                mid = (bids[0][0] + asks[0][0]) / 2
                
                print(f"\033[2J\033[H")
                print()
                print(f"{'=' * 70}")
                print(f"{'Exchange: BingX':^70}")
                print(f"{'GMT: ' + now_utc.strftime('%Y-%m-%d %H:%M:%S'):^70}")
                print(f"{'GMT+8: ' + now_gmt8.strftime('%Y-%m-%d %H:%M:%S'):^70}")
                print(f"{'Spread: $' + f'{spread:.2f}' + '  |  Mid: $' + f'{mid:,.2f}':^70}")
                print(f"{'=' * 70}")
                print()
                
                print(f"{'ASKS (Sell Orders)':^70}")
                print(f"{'Price':>11}  {'BTC':>10}  {'USDT':>9}  {'Cum BTC':>10}  {'Cum USDT':>9}")
                print(f"{'-' * 70}")
                
                cum_btc_ask = 0.0
                cum_usdt_ask = 0.0
                for p, v in reversed(asks):
                    if p > 0:
                        usdt = p * v
                        cum_btc_ask += v
                        cum_usdt_ask += usdt
                        print(f"\033[91m{p:>11,.2f}\033[0m  {v:>10,.7f}  {usdt:>9,.0f}  {cum_btc_ask:>10,.4f}  {cum_usdt_ask:>9,.0f}")
                    else:
                        print(f"{'---':>11}  {'---':>10}  {'---':>9}  {cum_btc_ask:>10,.4f}  {cum_usdt_ask:>9,.0f}")
                
                print()
                print(f"{'BIDS (Buy Orders)':^70}")
                print(f"{'Price':>11}  {'BTC':>10}  {'USDT':>9}  {'Cum BTC':>10}  {'Cum USDT':>9}")
                print(f"{'-' * 70}")
                
                cum_btc_bid = 0.0
                cum_usdt_bid = 0.0
                for p, v in bids:
                    if p > 0:
                        usdt = p * v
                        cum_btc_bid += v
                        cum_usdt_bid += usdt
                        print(f"\033[92m{p:>11,.2f}\033[0m  {v:>10,.7f}  {usdt:>9,.0f}  {cum_btc_bid:>10,.4f}  {cum_usdt_bid:>9,.0f}")
                    else:
                        print(f"{'---':>11}  {'---':>10}  {'---':>9}  {cum_btc_bid:>10,.4f}  {cum_usdt_bid:>9,.0f}")
                
                print(f"{'=' * 70}")
                
    except KeyboardInterrupt:
        feed.stop()
        print("\n✓ Stopped")

File: Crypto_Exchange/Orderbook/cpp/6_orderbook_bingx.cpp
#include "6_orderbook_bingx.h"
#include <boost/beast/core.hpp>
#include <boost/beast/ssl.hpp>
#include <boost/beast/websocket.hpp>
#include <boost/beast/websocket/ssl.hpp>
#include <boost/asio/connect.hpp>
#include <boost/asio/ip/tcp.hpp>
#include <boost/asio/ssl/stream.hpp>
#include <boost/iostreams/filtering_streambuf.hpp>
#include <boost/iostreams/copy.hpp>
#include <boost/iostreams/filter/gzip.hpp>
#include <iostream>
#include <iomanip>
#include <sstream>
#include <algorithm>
#include <chrono>
#include <ctime>
#include <mutex>

namespace beast = boost::beast;
namespace http = beast::http;
namespace websocket = beast::websocket;
namespace net = boost::asio;
namespace ssl = boost::asio::ssl;
using tcp = boost::asio::ip::tcp;

namespace bingx {

OrderBook::OrderBook() = default;
OrderBook::~OrderBook() { stop(); }

void OrderBook::start() {
    running_ = true;
    thread_ = std::thread([this] { run_loop(); });
}

void OrderBook::stop() {
    running_ = false;
    if (thread_.joinable()) thread_.join();
}

auto OrderBook::get_bids() const -> std::vector<Level> {
    std::lock_guard lock(mutex_);
    return bids_;
}

auto OrderBook::get_asks() const -> std::vector<Level> {
    std::lock_guard lock(mutex_);
    return asks_;
}

void OrderBook::process_message(const std::string& msg) {
    auto parse_levels = [](const std::string& text, size_t start) -> std::vector<Level> {
        std::vector<Level> levels;
        size_t pos = start;
        while (pos < text.size() && text[pos] != ']') {
            if (text[pos] == '[') {
                size_t end = text.find(']', pos);
                if (end == std::string::npos) break;
                std::string pair = text.substr(pos + 1, end - pos - 1);
                
                // Remove quotes: ["69660.83","0.002715"] -> 69660.83,0.002715
                std::string cleaned;
                for (char c : pair) {
                    if (c != '"') cleaned += c;
                }
                
                size_t comma = cleaned.find(',');
                if (comma != std::string::npos) {
                    try {
                        double price = std::stod(cleaned.substr(0, comma));
                        double volume = std::stod(cleaned.substr(comma + 1));
                        if (volume > 0.0) {
                            levels.push_back({price, volume});
                        }
                    } catch (...) {}
                }
                pos = end + 1;
            } else {
                ++pos;
            }
        }
        return levels;
    };
    
    auto bids_pos = msg.find("\"bids\"");
    auto asks_pos = msg.find("\"asks\"");
    
    if (bids_pos == std::string::npos || asks_pos == std::string::npos) return;
    
    auto bids_start = msg.find('[', bids_pos);
    auto asks_start = msg.find('[', asks_pos);
    
    if (bids_start == std::string::npos || asks_start == std::string::npos) return;
    
    auto bids = parse_levels(msg, bids_start + 1);
    auto asks = parse_levels(msg, asks_start + 1);
    
    std::sort(bids.begin(), bids.end(), [](auto& a, auto& b) { return a.price > b.price; });
    std::sort(asks.begin(), asks.end(), [](auto& a, auto& b) { return a.price < b.price; });
    
    if (!bids.empty() && !asks.empty()) {
        bids.erase(std::remove_if(bids.begin(), bids.end(), 
            [&](auto& b) { return b.price >= asks[0].price; }), bids.end());
        asks.erase(std::remove_if(asks.begin(), asks.end(), 
            [&](auto& a) { return a.price <= (bids.empty() ? 0.0 : bids[0].price); }), asks.end());
    }
    
    if (!bids.empty() && !asks.empty()) {
        std::lock_guard lock(mutex_);
        bids_ = std::move(bids);
        asks_ = std::move(asks);
    }
}

void OrderBook::run_loop() {
    while (running_) {
        try {
            net::io_context ioc;
            ssl::context ctx{ssl::context::tlsv12_client};
            ctx.set_default_verify_paths();
            
            tcp::resolver resolver{ioc};
            websocket::stream<beast::ssl_stream<tcp::socket>> ws{ioc, ctx};
            
            auto const results = resolver.resolve("open-api-ws.bingx.com", "443");
            auto ep = net::connect(get_lowest_layer(ws), results);
            
            if (!SSL_set_tlsext_host_name(ws.next_layer().native_handle(), "open-api-ws.bingx.com")) {
                throw beast::system_error(beast::error_code(static_cast<int>(::ERR_get_error()),
                    net::error::get_ssl_category()), "Failed to set SNI");
            }
            
            ws.next_layer().handshake(ssl::stream_base::client);
            ws.set_option(websocket::stream_base::decorator([](websocket::request_type& req) {
                req.set(http::field::user_agent, "BingX-OrderBook-C++20");
            }));
            
            ws.handshake("open-api-ws.bingx.com", "/market");
            
            std::string subscribe = R"({"id":"bingx-cpp-orderbook","reqType":"sub","dataType":"BTC-USDT@depth20"})";
            ws.write(net::buffer(subscribe));
            
            std::cout << "✓ BingX connected\n";
            
            while (running_) {
                beast::flat_buffer buffer;
                ws.read(buffer);
                
                std::string decompressed;
                try {
                    std::string compressed(static_cast<const char*>(buffer.data().data()), buffer.size());
                    std::istringstream iss(compressed);
                    boost::iostreams::filtering_streambuf<boost::iostreams::input> in;
                    in.push(boost::iostreams::gzip_decompressor());
                    in.push(iss);
                    std::ostringstream oss;
                    boost::iostreams::copy(in, oss);
                    decompressed = oss.str();
                } catch (...) {
                    decompressed = std::string(static_cast<const char*>(buffer.data().data()), buffer.size());
                }
                
                if (decompressed.find("\"ping\"") != std::string::npos) {
                    size_t ping_pos = decompressed.find("\"ping\"");
                    size_t colon = decompressed.find(':', ping_pos);
                    size_t end = decompressed.find_first_of(",}", colon);
                    if (colon != std::string::npos && end != std::string::npos) {
                        std::string ping_val = decompressed.substr(colon + 1, end - colon - 1);
                        std::string pong = "{\"pong\":" + ping_val + "}";
                        ws.write(net::buffer(pong));
                    }
                    continue;
                }
                
                if (decompressed.find("\"data\"") != std::string::npos) {
                    process_message(decompressed);
                }
            }
            
        } catch (std::exception const& e) {
            if (running_) {
                std::cerr << "BingX error: " << e.what() << ", reconnecting...\n";
                std::this_thread::sleep_for(std::chrono::seconds(2));
            }
        }
    }
}

} // namespace bingx

#ifndef NO_MAIN
int main() {
    bingx::OrderBook orderbook;
    orderbook.start();
    
    constexpr int DEPTH = 16;
    
    while (true) {
        std::this_thread::sleep_for(std::chrono::milliseconds(500));
        
        auto bids = orderbook.get_bids();
        auto asks = orderbook.get_asks();
        
        if (bids.empty() || asks.empty()) continue;
        
        bids.resize(std::min(bids.size(), size_t(DEPTH)));
        asks.resize(std::min(asks.size(), size_t(DEPTH)));
        
        while (bids.size() < DEPTH) bids.push_back({0.0, 0.0});
        while (asks.size() < DEPTH) asks.push_back({0.0, 0.0});
        
        double spread = asks[0].price - bids[0].price;
        double mid = (bids[0].price + asks[0].price) / 2.0;
        
        auto now = std::chrono::system_clock::now();
        auto now_t = std::chrono::system_clock::to_time_t(now);
        auto gmt = std::gmtime(&now_t);
        auto gmt8_t = now_t + 8 * 3600;
        auto gmt8 = std::gmtime(&gmt8_t);
        
        std::cout << "\033[2J\033[H\n";
        std::cout << "======================================================================\n";
        std::cout << "                          Exchange: BingX                             \n";
        std::cout << "                     GMT: " << std::put_time(gmt, "%Y-%m-%d %H:%M:%S") << "                     \n";
        std::cout << "                   GMT+8: " << std::put_time(gmt8, "%Y-%m-%d %H:%M:%S") << "                   \n";
        std::cout << "              Spread: $" << std::fixed << std::setprecision(2) << spread 
                  << "  |  Mid: $" << std::setprecision(2) << mid << "              \n";
        std::cout << "======================================================================\n\n";
        
        std::cout << "                        ASKS (Sell Orders)                            \n";
        std::cout << "      Price         BTC       USDT    Cum BTC   Cum USDT\n";
        std::cout << "----------------------------------------------------------------------\n";
        
        double cum_btc_ask = 0.0, cum_usdt_ask = 0.0;
        for (auto it = asks.rbegin(); it != asks.rend(); ++it) {
            if (it->price > 0) {
                double usdt = it->price * it->volume;
                cum_btc_ask += it->volume;
                cum_usdt_ask += usdt;
                std::cout << "\033[91m" << std::setw(11) << std::fixed << std::setprecision(2) << it->price << "\033[0m  "
                          << std::setw(10) << std::setprecision(7) << it->volume << "  "
                          << std::setw(9) << std::setprecision(0) << usdt << "  "
                          << std::setw(10) << std::setprecision(4) << cum_btc_ask << "  "
                          << std::setw(9) << std::setprecision(0) << cum_usdt_ask << "\n";
            } else {
                std::cout << std::setw(11) << "---" << "  "
                          << std::setw(10) << "---" << "  "
                          << std::setw(9) << "---" << "  "
                          << std::setw(10) << std::fixed << std::setprecision(4) << cum_btc_ask << "  "
                          << std::setw(9) << std::setprecision(0) << cum_usdt_ask << "\n";
            }
        }
        
        std::cout << "\n                        BIDS (Buy Orders)                             \n";
        std::cout << "      Price         BTC       USDT    Cum BTC   Cum USDT\n";
        std::cout << "----------------------------------------------------------------------\n";
        
        double cum_btc_bid = 0.0, cum_usdt_bid = 0.0;
        for (const auto& bid : bids) {
            if (bid.price > 0) {
                double usdt = bid.price * bid.volume;
                cum_btc_bid += bid.volume;
                cum_usdt_bid += usdt;
                std::cout << "\033[92m" << std::setw(11) << std::fixed << std::setprecision(2) << bid.price << "\033[0m  "
                          << std::setw(10) << std::setprecision(7) << bid.volume << "  "
                          << std::setw(9) << std::setprecision(0) << usdt << "  "
                          << std::setw(10) << std::setprecision(4) << cum_btc_bid << "  "
                          << std::setw(9) << std::setprecision(0) << cum_usdt_bid << "\n";
            } else {
                std::cout << std::setw(11) << "---" << "  "
                          << std::setw(10) << "---" << "  "
                          << std::setw(9) << "---" << "  "
                          << std::setw(10) << std::fixed << std::setprecision(4) << cum_btc_bid << "  "
                          << std::setw(9) << std::setprecision(0) << cum_usdt_bid << "\n";
            }
        }
        
        std::cout << "======================================================================\n";
    }
    
    return 0;
}

#endif // NO_MAIN
